### PR TITLE
feat(new transform): Add initial `transaction` transform

### LINKED ISSUE #1437: New `flatten` (`denormalize`, `canonical`, or `transaction`) transform
The purpose of this transform is to flatten related events into a single event representing a single transaction. A transaction is defined by a set of join factors, such as all events with the same `request_id` value. (see example below)

## Motivation

An increasingly popular tactic for observing applications is the use of an event stream. This is not a new concept, logs are effectively an event stream, and the idea of logs as events has been outlined in rather old guides (such as [12 factor](https://12factor.net/logs)). This was also the basis for the original Timber hosted service. The problem in the past was the impractical nature of storing this data. Storage solutions did not exist to properly analyze this data. That is changing and the use of rich event streams is becoming more popular. A common tactic to further optimize the performance of analyzing this data is to flatten or denormalize multiple events into a single "transaction" event.

## The problem

This is an excellent use case for Vector because users should not be denormalizing data in their application with the use of special tools, like vendor-specific libraries. This just restarts the lock-in cycle. Instead, users should be emitting normalized events from their app and then transforming that data with a tool like Vector. This decouples their app from whatever downstream system they're using.

## Example

For example, let's take a common Rails app that emits the following logs (events):

```js
{"timestamp": "...", "message": "Received GET /path", "request_id": "abcd1234", "request_path": "/path", "request_params": "..."}
{"timestamp": "...", "message": "Executed query in 5.2ms", "request_id": "abcd1234", "query": "SELECT * FROM table", "query_duration_ms": 5.2}
{"timestamp": "...", "message": "Rendered partial _partial.erb in 2.3ms", "request_id": "abcd1234", "template": "_partial.erb", "render_duration_ms": 2.3}
{"timestamp": "...", "message": "Executed query in 7.8ms", "request_id": "abcd1234", "query": "SELECT * FROM table", "query_duration_ms": 7.8}
{"timestamp": "...", "message": "Sent 200 in 15.2ms", "request_id": "abcd1234", "response_status": 200, "response_duration_ms": 5.2}
```

This transform would reduce (flatten) these events on the `request_id` field:

```json
{
  "timestamp_start": "...",
  "timestamp_end": "...", 
  "request_id": "abcd1234",
  "request_path": "/path",
  "request_params": "...",
  "query_duration_ms": 13.0,
  "render_duration_ms": 2.3,
  "status": 200,
 "response_duration_ms": 5.2
}
```

A few things to notice here:

1. The timestamp range represents the first and last event (obviously).
2. Duplicate string fields are discarded (`message`, `query`, `template`).
3. Unique string fields are kept (`params`).
3. Numeric fields are aggregated (`query_duration_ms`).

## Final thoughts

There's a lot of details left to discuss here, but it sets the stage for some of our marketing efforts around how to observe your application with Vector. Emitting normalized events decouples your app from any specific downstream solution which offers a plethora of benefits (no lock-in, clean backups, etc).

## Prior Art

- https://brandur.org/canonical-log-lines

## Marketing

- Normalized events streams
- Stripe's canonical log lines (https://stripe.com/blog/canonical-log-lines)

### PR DESCRIPTION
**Pull request by Jeffail**
This PR adds a new `transaction` transform. In order to flush events periodically (and one final flush during shutdown) it implements `transform_stream`.

Closes #1437

### CODE REVIEW

File: src/transforms/transaction.rs [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+struct TransactionState {
+    fields: HashMap<Atom, Box<dyn TransactionValueMerger>>,
+    stale_since: Instant,
+}
+
+impl TransactionState {
+    fn new(e: LogEvent, strategies: &IndexMap<Atom, MergeStrategy>) -> Self {
+        Self {
+            stale_since: Instant::now(),
+            // TODO: all_fields alternative that consumes
```
Comment (Jeffail) [inline src/transforms/transaction.rs]: Being forced to clone below which can be removed if we add a consuming version of `all_fields`.

**Comment by Jeffail** [comment]
Polling works as documented now thanks to compat layer so ready for review. Added a comment which in my view isn't a killer, but would improve things. Also haven't bothered fixing the `component-features` check until we're sure it's sticking around.

File: src/transforms/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+                        // Otherwise, it has been long enough since the last
+                        // flush to trigger another.
+                        last_flush = now;
+                        Poll::Ready(Some(Ok(StreamEvent::Flush)))
+                    }
+                }
+                // Pass `Poll<Option<Result<Event>>>` as `Poll<Option<Result<StreamEvent, ()>>>`
+                _ => return p.map(|p| p.map(|p| p.map(|p| StreamEvent::Event(p)))),
+            }
+        };
```
Comment (lukesteensen) [inline src/transforms/transaction.rs]: We should be able to structure this in a simpler way. If I'm understanding correctly, we have two "triggers": incoming events and flush/expiration timers. Incoming events are already modeled as a stream, and we should be able to use a tokio `DelayQueue` as a stream of flush/expirations (with the additional benefit of keys, so we don't have to iterate and check our whole state in `flush_into`).

With both of those streams available, we should be able to "just" loop and `select` over them and do the appropriate flush/flush all/insert action directly (i.e. no need for the `StreamEvent` intermediary). We may need to use [`async-stream`](https://github.com/tokio-rs/async-stream) for that (/cc @LucioFranco) and then compat the resulting stream before returning it.

If that all works out (:crossed_fingers:), we can get rid of the thread spawn, most of the manual plumbing, the full state iteration, etc.

**Review by lukesteensen — CHANGES_REQUESTED** [review]
This is looking awesome so far! Before merging I think we need to fix up the stream implementation, and it'd also be nice to see some more tests around the individual value mergers. Nothing crazy, but there are definitely some edge cases there I want to make sure we're exercising.

File: src/transforms/transaction.rs [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum MergeStrategy {
+    Discard,
+    Sum,
+    Max,
+    Min,
+    Array,
+    Append,
```
[the remark this answers is not in the document]
Comment (Jeffail) [reply src/transforms/transaction.rs]: That sounds good, concat makes more sense to me.

File: src/transforms/transaction.rs (hunk 2 above)
Comment (Jeffail) [reply src/transforms/transaction.rs]: I had a brief play around with creating a flush stream but I backed out as I couldn't see a simple and clean way of closing it once our event stream shuts down (without needing to return futures from `transform_stream`).

Not sure I'll have time this week to experiment.

**Comment by binarylogic** [comment]
@lukesteensen @Jeffail where do we sit with the stream changes? Are these absolutely necessary? @Jeffail is transitioning to our other project, so I'd like to limit follow up changes to a day or less. Otherwise, it might make sense to have someone else take this to the finish line. It might also be good practice with the new streaming transform API.

File: src/transforms/transaction.rs (hunk 2 above)
Comment (LucioFranco) [reply src/transforms/transaction.rs]: Not sure if this is useful or not but @MOZGIII has been working on an expiring type of hashmap for the new file sink https://github.com/timberio/vector/blob/tokio-compat-file-sink/src/expiring_hash_map/mod.rs, this may be useful.

**Comment by Jeffail** [comment]
It'd be nice to not have to iterate the state of pending transactions for each event so I think it's worth exploring an alternative. However, my gut feeling is that joining a channel owned by the transform itself within `transform_stream` is going to send me on a war with the compiler, and would require some trickery to have it shut down cleanly, and I'd rather not lose a day to it right now.

Given the use case for this transform I'm not convinced that these changes are worth blocking it unless someone wants to try it out themselves. If there's a significant and noticeable performance hit from the `flush_into` iterations then we can always add another period check so that it doesn't run for _each_ input event.

**Comment by a-rodin** [comment]
This looks great!

Does it make sense to add (maybe as a follow-up) an option which would limit the maximum number of events in a single transaction? This would act as protection in case if the transactions are created from data coming from untrusted sources and the downstream sinks can't process transactions which are too large.

**Review by lukesteensen — COMMENTED** [review]
The only thing I might consider blocking with the stream stuff is the thread spawn and sleep. That could end up having some weird runtime effects I'd rather avoid. Could we do something similar by spawning a `Delay` task, or even better, just having one flush timer that we poll?

/cc @LucioFranco for ideas

**Comment by binarylogic** [comment]
From planning: @lukesteensen will be finishing this up.

**Comment by binarylogic** [comment]
@lukesteensen should we close this in favor of https://github.com/timberio/vector/pull/2870.

**Review by binarylogic — CHANGES_REQUESTED** [review]
This is my fault 😄 , but we should rename this transform for clarity. We'll be directing customers towards this pattern and using a term that aligns with industry terms will benefit us. I'm leaning towards two different names:

1. `canonical` - Already established by Stripe. Devs curious about this can google this term and see that it's an adopted pattern.
2. `denormalize` - This is the more correct term, which is also much more clear in it's purpose.

I vote for `denormalize` since I don't fully understand why the term "canonical" was used by Stripe. Unless I'm missing technical context, it doesn't accurately describe what's happening.

**Comment by Hoverbear** [comment]
Closing to avoid future confusion
